File: 20.valid-parentheses.js
/*
 * @lc app=leetcode id=20 lang=javascript
 *
 * [20] Valid Parentheses
 */

// @lc code=start
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function (s) {
  let stk = [];
  let lookup = { "(": ")", "[": "]", "{": "}" };
  for (p of s) {
    if (p in lookup) {
      stk.push(p);
    } else if (stk.length === 0 || lookup[stk.pop()] !== p) {
      return false;
    }
  }
  return stk.length === 0;
};
// @lc code=end

File: 20.valid-parentheses.py
#
# @lc app=leetcode id=20 lang=python3
#
# [20] Valid Parentheses
#

# Time:  O(n)
# Space: O(n)

# @lc code=start
class Solution:
    def isValid(self, s: str) -> bool:
        stack, lookup = [], {"(": ")", "[": "]", "{": "}"}
        for p in s:
            if p in lookup:
                stack.append(p)
            elif len(stack) == 0 or lookup[stack.pop()] != p:
                return False
        return len(stack) == 0


# @lc code=end
